pkg/controller/batch/batch_controller.go

```diff
@@ -0,0 +1,365 @@
+var _ reconcile.Reconciler = &reconcileBatch{}
+
+// reconcileBatch reconciles a Batch object
+type reconcileBatch struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+type batchResource struct {
+	instance    *v2.Batch
+	client      client.Client
+	scheme      *runtime.Scheme
+	requestName types.NamespacedName
+}
+
+func (r *reconcileBatch) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Batch")
+
+	// Fetch the Batch instance
+	instance := &v2.Batch{}
+	err := r.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	batch := &batchResource{
+		instance:    instance,
+		client:      r.Client,
+		scheme:      r.scheme,
+		requestName: request.NamespacedName,
+	}
+
+	phase := instance.Status.Phase
+	switch phase {
+	case "":
+		return reconcile.Result{}, batch.UpdatePhase(v2.BatchInitializing, nil)
+	case v2.BatchInitializing:
+		return batch.initializeResources()
+	case v2.BatchInitialized:
+		return batch.execute()
+	case v2.BatchRunning:
+		return batch.waitToComplete()
+	}
+	// Backup either succeeded or failed, so cleanup resources
+	return batch.cleanup()
+}
+
+func (r *batchResource) initializeResources() (reconcile.Result, error) {
+	batch := r.instance
+	spec := batch.Spec
+	// Ensure the Infinispan cluster exists
+	infinispan := &v1.Infinispan{}
+	if result, err := kube.LookupResource(spec.Cluster, batch.Namespace, infinispan, r.client, log); result != nil {
+		return *result, err
+	}
+
+	if err := infinispan.EnsureClusterStability(); err != nil {
+		log.Info(fmt.Sprintf("Infinispan '%s' not ready: %s", spec.Cluster, err.Error()))
+		return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, nil
+	}
+
+	if spec.ConfigMap == nil && spec.Config == nil {
+		return reconcile.Result{},
+			r.UpdatePhase(v2.BatchFailed, fmt.Errorf("At most one of ['Spec.config', 'spec.ConfigMap'] must be configured"))
+	}
+
+	if spec.ConfigMap == nil {
+		// Create configMap
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      batch.Name,
+				Namespace: batch.Namespace,
+			},
+		}
+		_, err := controllerutil.CreateOrUpdate(ctx, r.client, configMap, func() error {
+			configMap.Data = map[string]string{BatchFilename: *spec.Config}
+			return controllerutil.SetControllerReference(batch, configMap, r.scheme)
+		})
+
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Unable to create ConfigMap '%s': %w", configMap.Name, err)
+		}
+
+		updated, err := r.update(func() {
```

Comment:
As Batch CR one-time operation, we should set Infinispan CR as Batch CR controller OwnerReference